File: app/src/main/java/org/gittner/osmbugs/api/MapdustApi.java
package org.gittner.osmbugs.api;

import android.util.Log;

import org.gittner.osmbugs.bugs.MapdustBug;
import org.gittner.osmbugs.common.Comment;
import org.gittner.osmbugs.parser.MapdustParser;
import org.gittner.osmbugs.statics.Settings;
import org.osmdroid.util.BoundingBoxE6;
import org.osmdroid.util.GeoPoint;

import java.io.IOException;
import java.net.ProtocolException;
import java.util.ArrayList;
import java.util.Locale;

import okhttp3.FormBody;
import okhttp3.HttpUrl;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class MapdustApi implements BugApi<MapdustBug>
{
    private static final String API_KEY = "ae58b0b4aa3f876265a4d5f29167b73c";

    public static OkHttpClient mOkHttpClient = new OkHttpClient();

    public ArrayList<MapdustBug> downloadBBox(BoundingBoxE6 bBox)
    {
        Request request = new Request.Builder()
                .url(new HttpUrl.Builder()
                        .scheme("http")
                        .host(!Settings.isDebugEnabled() ? "www.mapdust.com" : "st.www.mapdust.com")
                        .addPathSegment("api")
                        .addPathSegment("getBugs")
                        .addQueryParameter("key", API_KEY)
                        .addQueryParameter("bbox", String.format(
                                Locale.US,
                                "%f,%f,%f,%f",
                                bBox.getLonEastE6() / 1000000.0,
                                bBox.getLatSouthE6() / 1000000.0,
                                bBox.getLonWestE6() / 1000000.0,
                                bBox.getLatNorthE6() / 1000000.0))
                        .addQueryParameter("comments", "1")
                        .addQueryParameter("ft", getMapdustSelectionString())
                        .addQueryParameter("fs", getMapdustEnabledTypesString())
                        .build())
                .build();

        try
        {
            Response response = mOkHttpClient.newCall(request).execute();

            if (response.code() != 200)
            {
                return null;
            }

            return MapdustParser.parse(response.body().byteStream());
        }
        catch (ProtocolException e)
        {
            if (e.getMessage().startsWith("HTTP 204"))
            {
                return new ArrayList<>();
            }
        }
        catch (IOException e)
        {
            e.printStackTrace();
        }

        return  null;
    }


    private String getMapdustSelectionString()
    {
        String result = "";
        if (Settings.Mapdust.isWrongTurnEnabled())
        {
            result += "wrong_turn,";
        }
        if (Settings.Mapdust.isBadRoutingEnabled())
        {
            result += "bad_routing,";
        }
        if (Settings.Mapdust.isOnewayRoadEnabled())
        {
            result += "oneway_road,";
        }
        if (Settings.Mapdust.isBlockedStreetEnabled())
        {
            result += "blocked_street,";
        }
        if (Settings.Mapdust.isMissingStreetEnabled())
        {
            result += "missing_street,";
        }
        if (Settings.Mapdust.isRoundaboutIssueEnabled())
        {
            result += "wrong_roundabout,";
        }
        if (Settings.Mapdust.isMissingSpeedInfoEnabled())
        {
            result += "missing_speedlimit,";
        }
        if (Settings.Mapdust.isOtherEnabled())
        {
            result += "other,";
        }
        if (result.endsWith(","))
        {
            result = result.substring(0, result.length() - 1);
        }
        return result;
    }


    private String getMapdustEnabledTypesString()
    {
        String result = "";
        if (Settings.Mapdust.isShowOpenEnabled())
        {
            result += "1,";
        }
        if (Settings.Mapdust.isShowClosedEnabled())
        {
            result += "2,";
        }
        if (Settings.Mapdust.isShowIgnoredEnabled())
        {
            result += "3,";
        }
        if (result.endsWith(","))
        {
            result = result.substring(0, result.length() - 1);
        }
        return result;
    }


    public boolean commentBug(long id, String comment, String nickname)
    {
        Request request = new Request.Builder()
                .url(new HttpUrl.Builder()
                        .scheme("http")
                        .host(!Settings.isDebugEnabled() ? "www.mapdust.com" : "st.www.mapdust.com")
                        .addPathSegment("api")
                        .addPathSegment("commentBug")
                        .addQueryParameter("key", API_KEY)
                        .addQueryParameter("id", String.valueOf(id))
                        .addQueryParameter("comment", comment)
                        .addQueryParameter("nickname", nickname)
                        .build())
                .post(new FormBody.Builder().build())
                .build();

        try
        {
            Response response = mOkHttpClient.newCall(request).execute();

            /* Mapdust returns 201 for commentBug as Success */
            return response.code() == 201;
        }
        catch (IOException e)
        {
            e.printStackTrace();
        }

        return false;
    }


    public boolean changeBugStatus(long id, MapdustBug.STATE state, String comment, String username)
    {
        String sState = "";
        switch (state)
        {
            case OPEN:
                sState= "1";
                break;

            case CLOSED:
                sState = "2";
                break;

            default:
                sState = "3";
                break;
        }

        Request request = new Request.Builder()
                .url(new HttpUrl.Builder()
                        .scheme("http")
                        .host(!Settings.isDebugEnabled() ? "www.mapdust.com" : "st.www.mapdust.com")
                        .addPathSegment("api")
                        .addPathSegment("changeBugStatus")
                        .addQueryParameter("key", API_KEY)
                        .addQueryParameter("id", String.valueOf(id))
                        .addQueryParameter("comment", comment)
                        .addQueryParameter("nickname", username)
                        .addQueryParameter("status", sState)
                        .build())
                .post(new FormBody.Builder().build())
                .build();

        try
        {
            Response response = mOkHttpClient.newCall(request).execute();

            /* Mapdust returns 201 for changeBugStatus as Success */
            return response.code() == 201;
        }
        catch (IOException e)
        {
            e.printStackTrace();
        }

        return false;
    }


    public boolean addBug(GeoPoint position, int type, String description)
    {
        String sType = "";
        switch (type)
        {
            case MapdustBug.WRONG_TURN:
                sType = "wrong_turn";
                break;

            case MapdustBug.BAD_ROUTING:
                sType = "bad_routing";
                break;

            case MapdustBug.ONEWAY_ROAD:
                sType = "oneway_road";
                break;

            case MapdustBug.BLOCKED_STREET:
                sType = "blocked_street";
                break;

            case MapdustBug.MISSING_STREET:
                sType = "missing_street";
                break;

            case MapdustBug.ROUNDABOUT_ISSUE:
                sType = "wrong_roundabout";
                break;

            case MapdustBug.MISSING_SPEED_INFO:
                sType = "missing_speedlimit";
                break;

            case MapdustBug.OTHER:
                sType = "other";
                break;

            default:
                return false;
        }

        Request request = new Request.Builder()
                .url(new HttpUrl.Builder()
                        .scheme("http")
                        .host(!Settings.isDebugEnabled() ? "www.mapdust.com" : "st.www.mapdust.com")
                        .addPathSegment("api")
                        .addPathSegment("addBug")
                        .addQueryParameter("key", API_KEY)
                        .addQueryParameter("coordinates", String.format(
                                Locale.US,
                                "%f,%f",
                                position.getLongitudeE6() / 1000000.0,
                                position.getLatitudeE6() / 1000000.0))
                        .addQueryParameter("description", description)
                        .addQueryParameter("nickname", Settings.Mapdust.getUsername())
                        .addQueryParameter("type", sType)
                        .addQueryParameter("nickname", Settings.Mapdust.getUsername())
                        .build())
                .post(new FormBody.Builder().build())
                .build();

        try
        {
            Response response = mOkHttpClient.newCall(request).execute();

            /* Mapdust returns 201 for addBug as Success */
            return response.code() == 201;
        }
        catch (IOException e)
        {
            e.printStackTrace();
        }

        return false;
    }


    public ArrayList<Comment> retrieveComments(long id)
    {
        ArrayList<Comment> comments = new ArrayList<>();

        Request request = new Request.Builder()
                .url(new HttpUrl.Builder()
                        .scheme("http")
                        .host(!Settings.isDebugEnabled() ? "www.mapdust.com" : "st.www.mapdust.com")
                        .addPathSegment("api")
                        .addPathSegment("getBug")
                        .addQueryParameter("key", API_KEY)
                        .addQueryParameter("id", String.valueOf(id))
                        .build())
                .post(new FormBody.Builder().build())
                .build();

        try
        {
            Response response = mOkHttpClient.newCall(request).execute();

            if (response.code() != 200)
            {
                return comments;
            }

            /* If Request was Successful, parse the Stream */
            comments.addAll(MapdustParser.parseSingleBugForComments(response.body().byteStream()));
        }
        catch (IOException e)
        {
            e.printStackTrace();
        }

        return comments;
    }
}

File: app/src/main/java/org/gittner/osmbugs/mapdust/MapdustApi.kt
package org.gittner.osmbugs.mapdust

import android.net.Uri
import com.github.kittinunf.fuel.Fuel
import com.github.kittinunf.fuel.coroutines.awaitStringResponse
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.gittner.osmbugs.statics.Settings
import org.koin.core.KoinComponent
import org.koin.core.inject
import org.osmdroid.api.IGeoPoint
import org.osmdroid.util.BoundingBox
import java.util.*
import kotlin.collections.ArrayList


class MapdustApi : KoinComponent {
    private val API_KEY = "ae58b0b4aa3f876265a4d5f29167b73c"

    private val mSettings = Settings.getInstance()

    suspend fun download(bBox: BoundingBox): ArrayList<MapdustError> {
        val url = Uri.Builder()
            .scheme("http")
            .authority("www.mapdust.com")
            .appendPath("api")
            .appendPath("getBugs")
            .appendQueryParameter("key", API_KEY)
            .appendQueryParameter("bbox", "%f,%f,%f,%f".format(Locale.US, bBox.lonEast, bBox.latSouth, bBox.lonWest, bBox.latNorth))
            .appendQueryParameter("comments", "1")
            .appendQueryParameter("ft", getMapdustSelectionString())
            .appendQueryParameter("fs", getMapdustEnabledTypesString())
            .build()

        val response = Fuel
            .get(url.toString())
            .awaitStringResponse()

        // No content
        if (response.second.statusCode == 204) {
            return ArrayList()
        }
        if (response.second.statusCode != 200) {
            throw RuntimeException("Invalid status code: ${response.second.statusCode}")
        }

        return MapdustParser.parse(response.third)
    }

    private fun getMapdustSelectionString(): String? {
        var result = ""
        if (mSettings.Mapdust.GetTypeEnabled(MapdustError.ERROR_TYPE.WRONG_TURN)) {
            result += "wrong_turn,"
        }
        if (mSettings.Mapdust.GetTypeEnabled(MapdustError.ERROR_TYPE.BAD_ROUTING)) {
            result += "bad_routing,"
        }
        if (mSettings.Mapdust.GetTypeEnabled(MapdustError.ERROR_TYPE.ONEWAY_ROAD)) {
            result += "oneway_road,"
        }
        if (mSettings.Mapdust.GetTypeEnabled(MapdustError.ERROR_TYPE.BLOCKED_STREET)) {
            result += "blocked_street,"
        }
        if (mSettings.Mapdust.GetTypeEnabled(MapdustError.ERROR_TYPE.MISSING_STREET)) {
            result += "missing_street,"
        }
        if (mSettings.Mapdust.GetTypeEnabled(MapdustError.ERROR_TYPE.ROUNDABOUT_ISSUE)) {
            result += "wrong_roundabout,"
        }
        if (mSettings.Mapdust.GetTypeEnabled(MapdustError.ERROR_TYPE.MISSING_SPEED_INFO)) {
            result += "missing_speedlimit,"
        }
        if (mSettings.Mapdust.GetTypeEnabled(MapdustError.ERROR_TYPE.OTHER)) {
            result += "other,"
        }
        if (result.endsWith(",")) {
            result = result.substring(0, result.length - 1)
        }
        return result
    }


    private fun getMapdustEnabledTypesString(): String? {
        var result = ""
        if (mSettings.Mapdust.ShowOpen) {
            result += "1,"
        }
        if (mSettings.Mapdust.ShowClosed) {
            result += "2,"
        }
        if (mSettings.Mapdust.ShowIgnored) {
            result += "3,"
        }
        if (result.endsWith(",")) {
            result = result.substring(0, result.length - 1)
        }
        return result
    }

    suspend fun download(id: Long): MapdustError = withContext(Dispatchers.IO) {
        val url = Uri.Builder()
            .scheme("http")
            .authority("www.mapdust.com")
            .appendPath("api")
            .appendPath("getBug")
            .appendQueryParameter("key", API_KEY)
            .appendQueryParameter("id", id.toString())
            .build()

        val response = Fuel
            .get(url.toString())
            .awaitStringResponse()

        if (response.second.statusCode != 200) {
            throw RuntimeException("Invalid status code: ${response.second.statusCode}")
        }

        MapdustParser.parseError(response.third)
    }

    suspend fun commentError(Id: Long, comment: String) = withContext(Dispatchers.IO) {
        val url = Uri.Builder()
            .scheme("http")
            .authority("www.mapdust.com")
            .appendPath("api")
            .appendPath("commentBug")
            .appendQueryParameter("key", API_KEY)
            .appendQueryParameter("id", Id.toString())
            .appendQueryParameter("comment", comment)
            .appendQueryParameter("nickname", mSettings.Mapdust.Username)
            .build()

        val response = Fuel
            .post(url.toString())
            .awaitStringResponse()

        // Mapdust returns 201 for commentBug as Success
        if (response.second.statusCode != 201) {
            throw RuntimeException("Invalid status code: ${response.second.statusCode}")
        }
    }

    suspend fun addBug(point: IGeoPoint, type: MapdustError.ERROR_TYPE, description: String): Long = withContext(Dispatchers.IO) {
        val sType = when (type) {
            MapdustError.ERROR_TYPE.WRONG_TURN -> "wrong_turn"
            MapdustError.ERROR_TYPE.BAD_ROUTING -> "bad_routing"
            MapdustError.ERROR_TYPE.ONEWAY_ROAD -> "oneway_road"
            MapdustError.ERROR_TYPE.BLOCKED_STREET -> "blocked_street"
            MapdustError.ERROR_TYPE.MISSING_STREET -> "missing_street"
            MapdustError.ERROR_TYPE.ROUNDABOUT_ISSUE -> "wrong_roundabout"
            MapdustError.ERROR_TYPE.MISSING_SPEED_INFO -> "missing_speedlimit"
            MapdustError.ERROR_TYPE.OTHER -> "other"
        }

        val url = Uri.Builder()
            .scheme("http")
            .authority("www.mapdust.com")
            .appendPath("api")
            .appendPath("addBug")
            .appendQueryParameter("key", API_KEY)
            .appendQueryParameter("coordinates", "%f,%f".format(Locale.ENGLISH, point.longitude, point.latitude))
            .appendQueryParameter("description", description)
            .appendQueryParameter("nickname", mSettings.Mapdust.Username)
            .appendQueryParameter("type", sType)
            .build()

        val response = Fuel
            .post(url.toString())
            .awaitStringResponse()

        // Mapdust returns 201 for addBug as Success
        if (response.second.statusCode != 201) {
            throw RuntimeException("Invalid status code: ${response.second.statusCode}")
        }

        MapdustParser.parseAddBug(response.third)
    }

    suspend fun changeErrorStatus(id: Long, comment: String, state: MapdustError.STATE) = withContext(Dispatchers.IO) {
        val sState = when (state) {
            MapdustError.STATE.OPEN -> "1"
            MapdustError.STATE.CLOSED -> "2"
            else -> "3"
        }

        val url = Uri.Builder()
            .scheme("http")
            .authority("www.mapdust.com")
            .appendPath("api")
            .appendPath("changeBugStatus")
            .appendQueryParameter("key", API_KEY)
            .appendQueryParameter("id", id.toString())
            .appendQueryParameter("comment", comment)
            .appendQueryParameter("nickname", mSettings.Mapdust.Username)
            .appendQueryParameter("status", sState)
            .build()

        val response = Fuel
            .post(url.toString())
            .awaitStringResponse()

        // Mapdust returns 201 for addBug as Success
        if (response.second.statusCode != 201) {
            throw RuntimeException("Invalid status code: ${response.second.statusCode}")
        }
    }
}